src/deadline/mcp/utils.py

```diff
@@ -0,0 +1,116 @@
+import json
+import logging
+from typing import Any, Callable, Dict, Optional
+
+from mcp.server.fastmcp import FastMCP
+
+from .config import get_all_tool_names, get_tool_config, ToolConfig
+
+logger = logging.getLogger(__name__)
+
+
+def _get_original_docstring(func: Callable) -> Optional[str]:
+    """Get the original docstring from a function, bypassing decorator wrappers."""
+    if hasattr(func, "__closure__") and func.__closure__:
+        for cell in func.__closure__:
+            try:
+                content = cell.cell_contents
+                # Skip telemetry wrapper doc strings
+                if (
+                    callable(content)
+                    and hasattr(content, "__doc__")
+                    and content.__doc__
+                    and content.__doc__.strip()
+                    and "Wrapper to time a function for latency telemetry" not in content.__doc__
+                ):
+                    return content.__doc__
+            except ValueError:
+                continue
+
+    return func.__doc__
+
+
+def _default_serializer(obj: Any) -> Any:
+    """Default serializer for API responses."""
+    if hasattr(obj, "__dict__"):
+        return obj.__dict__
+    elif hasattr(obj, "to_dict"):
+        return obj.to_dict()
+    return str(obj)
+
+
+def _default_error_handler(e: Exception) -> Dict:
+    """Default error handler for API calls."""
+    error_info = {"error": str(e), "type": type(e).__name__}
+    if hasattr(e, "response") and hasattr(e.response, "status_code"):
+        error_info["status_code"] = e.response.status_code
+    logger.error(f"API tool error: {error_info}", exc_info=True)
+    return error_info
+
+
+def _create_wrapper(config: ToolConfig, serializer: Callable, error_handler: Callable) -> Callable:
+    """Create a wrapper function based on the tool configuration."""
+    func = config["func"]
+    params = config["params"]
+
+    if params is None:
+        sig = inspect.signature(func)
+        params = [p for p in sig.parameters.keys() if p != "config"]
+
+    signature = inspect.Signature(
+        [
+            inspect.Parameter(name, inspect.Parameter.KEYWORD_ONLY, default=None, annotation=str)
+            for name in params
+        ],
+        return_annotation=dict,
+    )
+
+    def wrapper(**kwargs) -> dict:
+        try:
+            # Filter out empty/null values that MCP clients might send
+            filtered_kwargs = {k: v for k, v in kwargs.items() if v not in (None, "", "null")}
+            # Deadline Cloud API functions accept an optional 'config' parameter
+            # for custom ConfigParser objects. In MCP tools, we always pass None
+            # to use the default configuration (from ~/.deadline/config or environment).
+            if "config" in inspect.signature(func).parameters:
+                filtered_kwargs["config"] = None
+
+            result = func(**filtered_kwargs)
+            return json.loads(json.dumps(result, default=serializer))
```

Comment:
Why do we do this? Could use a comment if it's intentional. I'd guess so the MCP server can serialize it with the default serializer later but I'm not sure.